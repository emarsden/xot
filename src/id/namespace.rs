use super::idmap::{IdIndex, IdMap};

/// Id uniquely identifying namespace.
#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, Ord, PartialOrd)]
pub struct NamespaceId(u16);

impl IdIndex<NamespaceId> for NamespaceId {
    fn to_id(index: usize) -> NamespaceId {
        NamespaceId(index as u16)
    }

    fn from_id(id: NamespaceId) -> usize {
        id.0 as usize
    }
}

pub(crate) type NamespaceLookup = IdMap<NamespaceId, String>;
