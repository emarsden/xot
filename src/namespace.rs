use std::fmt::{Display, Formatter};

use crate::idmap::{IdIndex, IdMap};

/// Id uniquely identifying namespace.
#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, Ord, PartialOrd)]
pub struct NamespaceId(u16);

impl IdIndex<NamespaceId> for NamespaceId {
    fn to_id(index: usize) -> NamespaceId {
        NamespaceId(index as u16)
    }

    fn from_id(id: NamespaceId) -> usize {
        id.0 as usize
    }
}

#[derive(Debug, Clone, Hash, PartialEq, Eq)]
pub struct Namespace(String);

impl Namespace {
    pub(crate) fn new(namespace_uri: String) -> Self {
        Self(namespace_uri)
    }

    pub(crate) fn get(&self) -> &str {
        &self.0
    }
}

impl Display for Namespace {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

pub(crate) type NamespaceLookup = IdMap<NamespaceId, Namespace>;
