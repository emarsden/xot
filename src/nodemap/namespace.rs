use crate::xmlvalue::{Namespace, Value, ValueCategory};
use crate::{NamespaceId, Node, PrefixId, Xot};

use super::core::{category_predicate, NodeMap, ValueAdapter};

pub struct NamespaceAdapter {}

impl ValueAdapter<PrefixId, NamespaceId> for NamespaceAdapter {
    fn children(xot: &Xot, node: Node) -> impl Iterator<Item = Node> + '_ {
        xot.all_children(node)
            .take_while(category_predicate(xot, ValueCategory::Namespace))
    }

    fn insertion_point(xot: &Xot, node: Node) -> Option<Node> {
        let last_child = Self::children(xot, node).last();
        // if there is a last child, insert after it
        if let Some(last_child) = last_child {
            return Some(last_child);
        }
        // if there is no namespace node, we want to prepend
        None
    }

    fn key(value: &Value) -> &PrefixId {
        match value {
            Value::Namespace(Namespace { prefix_id, .. }) => prefix_id,
            _ => unreachable!(),
        }
    }

    fn value(value: &Value) -> &NamespaceId {
        match value {
            Value::Namespace(Namespace { namespace_id, .. }) => namespace_id,
            _ => unreachable!(),
        }
    }

    fn value_mut(value: &mut Value) -> &mut NamespaceId {
        match value {
            Value::Namespace(Namespace { namespace_id, .. }) => namespace_id,
            _ => unreachable!(),
        }
    }

    fn create(key: PrefixId, value: NamespaceId) -> Value {
        Value::Namespace(Namespace {
            prefix_id: key,
            namespace_id: value,
        })
    }

    fn update(value: &mut Value, new_value: NamespaceId) -> Option<NamespaceId> {
        match value {
            Value::Namespace(Namespace {
                namespace_id: old_value,
                ..
            }) => {
                let old_value = std::mem::replace(old_value, new_value);
                Some(old_value)
            }
            _ => unreachable!(),
        }
    }
}

pub type Namespaces<'a> = NodeMap<'a, PrefixId, NamespaceId, NamespaceAdapter>;
