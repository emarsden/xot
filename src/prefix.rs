use std::borrow::Cow;
use std::fmt::{Display, Formatter};

use crate::idmap::{IdIndex, IdMap};

#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
pub struct PrefixId(u16);

impl IdIndex<PrefixId> for PrefixId {
    fn to_id(index: usize) -> PrefixId {
        PrefixId(index as u16)
    }

    fn from_id(id: PrefixId) -> usize {
        id.0 as usize
    }
}

#[derive(Debug, Clone, Hash, PartialEq, Eq)]
pub(crate) struct Prefix<'a>(Cow<'a, str>);

impl<'a> Prefix<'a> {
    pub(crate) fn new(prefix: &'a str) -> Self {
        Self(prefix.into())
    }
}

impl<'a> Display for Prefix<'a> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

pub(crate) type PrefixLookup<'a> = IdMap<PrefixId, Prefix<'a>>;
