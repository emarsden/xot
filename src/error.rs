use crate::namespace::Namespace;

#[derive(Debug)]
pub enum Error {
    NoPrefixForNamespace(String),
    UnknownPrefix(String),
    Io(std::io::Error),
    NodeId(id_tree::NodeIdError),
    Parser(xmlparser::Error),
}

impl From<xmlparser::Error> for Error {
    #[inline]
    fn from(e: xmlparser::Error) -> Self {
        Error::Parser(e)
    }
}

impl From<id_tree::NodeIdError> for Error {
    #[inline]
    fn from(e: id_tree::NodeIdError) -> Self {
        Error::NodeId(e)
    }
}

impl From<std::io::Error> for Error {
    #[inline]
    fn from(e: std::io::Error) -> Self {
        Error::Io(e)
    }
}
