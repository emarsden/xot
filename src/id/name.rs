use super::idmap::{IdIndex, IdMap};
use super::namespace::NamespaceId;

/// Id uniquely identifying a name and namespace.
#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, Ord, PartialOrd)]
pub struct NameId(pub(crate) u16);

impl IdIndex<NameId> for NameId {
    fn to_id(index: usize) -> NameId {
        NameId(index as u16)
    }

    fn from_id(id: NameId) -> usize {
        id.0 as usize
    }
}

#[derive(Debug, Clone, Hash, PartialEq, Eq)]
pub(crate) struct Name {
    pub(crate) name: String,
    pub(crate) namespace_id: NamespaceId,
}

impl Name {
    pub(crate) fn new<S: Into<String>>(name: S, namespace_id: NamespaceId) -> Self {
        Self {
            name: name.into(),
            namespace_id,
        }
    }
}

pub(crate) type NameLookup = IdMap<NameId, Name>;
