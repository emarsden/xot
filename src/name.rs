use std::borrow::Cow;

use crate::idmap::{IdIndex, IdMap};
use crate::namespace::NamespaceId;

/// Id uniquely identifying a name and namespace.
#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, Ord, PartialOrd)]
pub struct NameId(u16);

impl IdIndex<NameId> for NameId {
    fn to_id(index: usize) -> NameId {
        NameId(index as u16)
    }

    fn from_id(id: NameId) -> usize {
        id.0 as usize
    }
}

#[derive(Debug, Clone, Hash, PartialEq, Eq)]
pub(crate) struct Name<'a> {
    pub(crate) name: Cow<'a, str>,
    pub(crate) namespace_id: NamespaceId,
}

impl<'a> Name<'a> {
    pub(crate) fn new<S: Into<Cow<'a, str>>>(name: S, namespace_id: NamespaceId) -> Self {
        Self {
            name: name.into(),
            namespace_id,
        }
    }
}

pub(crate) type NameLookup<'a> = IdMap<NameId, Name<'a>>;
