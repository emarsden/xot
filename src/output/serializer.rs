use genawaiter::rc::gen;
use genawaiter::yield_;

use crate::access::NodeEdge;
use crate::id::{NameId, NamespaceId, PrefixId};
use crate::xmlvalue::{Element, Prefixes};
use crate::xmlvalue::{Value, ValueType};
use crate::xotdata::{Node, Xot};

/// Output of serialization
///
/// Given an [`OutputToken`] or
/// [`PrettyOutputToken`](`crate::output::PrettyOutputToken`), this enum
/// represents what the token represents in the XML tree.
///
/// You can use this information for customized serialization.
#[derive(Debug, PartialEq)]
pub enum Output<'a> {
    /// Start tag open, i.e `<foo` or `<ns:foo`
    StartTagOpen(Element),
    /// Start tag close, either `>` or `/>`
    StartTagClose,
    /// End tag, i.e. `</foo>` or `</ns:foo>`
    EndTag(Element),
    /// Namespace declaration, i.e. `xmlns:foo="http://example.com"`
    Prefix(PrefixId, NamespaceId),
    /// Attribute, i.e. `foo="bar"`
    Attribute(NameId, &'a str),
    /// Text, i.e. `foo`
    Text(&'a str),
    /// Comment, i.e. `<!-- foo -->`
    Comment(&'a str),
    /// Processing instruction, i.e. `<?foo bar?>`
    ProcessingInstruction(NameId, Option<&'a str>),
}

pub(crate) fn gen_outputs(xot: &Xot, node: Node) -> impl Iterator<Item = (Node, Output)> + '_ {
    gen!({
        for edge in xot.traverse(node) {
            match edge {
                NodeEdge::Start(current_node) => {
                    let gen = gen_edge_start(xot, node, current_node);
                    for output in gen {
                        yield_!((current_node, output));
                    }
                }
                NodeEdge::End(current_node) => {
                    let gen = gen_edge_end(xot, current_node);
                    for output in gen {
                        yield_!((current_node, output));
                    }
                }
            }
        }
    })
    .into_iter()
}

fn gen_edge_start(xot: &Xot, top_node: Node, node: Node) -> impl Iterator<Item = Output> + '_ {
    gen!({
        let value = xot.value(node);

        match value {
            Value::Document => {}
            Value::Element(element) => {
                yield_!(Output::StartTagOpen(*element));

                // serialize any extra prefixes if this is the top element of
                // a fragment and they aren't declared already
                let namespaces = xot.namespaces(node);
                if node == top_node {
                    for (prefix_id, namespace_id) in xot.namespaces_in_scope(node) {
                        if !namespaces.contains_key(prefix_id) {
                            yield_!(Output::Prefix(prefix_id, namespace_id,));
                        }
                    }
                }

                for (prefix_id, namespace_id) in namespaces.iter() {
                    yield_!(Output::Prefix(prefix_id, *namespace_id,));
                }

                for (name_id, value) in xot.attributes(node).iter() {
                    yield_!(Output::Attribute(name_id, value));
                }

                yield_!(Output::StartTagClose);
            }
            Value::Text(text) => {
                yield_!(Output::Text(text.get()));
            }
            Value::Comment(comment) => {
                yield_!(Output::Comment(comment.get()));
            }
            Value::ProcessingInstruction(pi) => {
                yield_!(Output::ProcessingInstruction(pi.target(), pi.data()));
            }
            Value::Attribute(_) | Value::Namespace(_) => {
                // handled in element
            }
        }
    })
    .into_iter()
}

fn gen_edge_end(xot: &Xot, node: Node) -> impl Iterator<Item = Output> + '_ {
    gen!({
        let value = xot.value(node);
        if let Value::Element(element) = value {
            yield_!(Output::EndTag(*element));
        }
    })
    .into_iter()
}

/// Output token
///
/// This represents an [`Output`] as a rendered output token.
pub struct OutputToken {
    /// Whether the token is prefixed by a space character.
    pub space: bool,
    /// The token.
    ///
    /// This is a fragment of XML like `<foo` or `a="A"` or `/>`, etc.
    pub text: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_iter_mkgen() {
        let mut xot = Xot::new();
        let root = xot.parse(r#"<doc a="A">Text</doc>"#).unwrap();
        let a_id = xot.add_name("a");
        let doc = xot.document_element(root).unwrap();
        let doc_el = xot.element(doc).unwrap();
        let mut iter = gen_outputs(&xot, doc);

        let v = iter.next().unwrap().1;
        assert_eq!(v, Output::StartTagOpen(*doc_el));
        let v = iter.next().unwrap().1;
        assert_eq!(v, Output::Prefix(xot.xml_prefix(), xot.xml_namespace()));
        let v = iter.next().unwrap().1;
        assert_eq!(v, Output::Attribute(a_id, "A"));
        let v = iter.next().unwrap().1;
        assert_eq!(v, Output::StartTagClose);
        let v = iter.next().unwrap().1;
        assert_eq!(v, Output::Text("Text"));
        let v = iter.next().unwrap().1;
        assert_eq!(v, Output::EndTag(*doc_el));
        assert!(iter.next().is_none());
    }
}
