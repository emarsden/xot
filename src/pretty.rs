use std::io::Write;

use crate::error::Error;
use crate::name::NameId;
use crate::namespace::NamespaceId;
use crate::prefix::PrefixId;
use crate::serializer::{SerializerWriter, XmlSerializerWriter};
use crate::xmlvalue::{Comment, Element, ProcessingInstruction, Text, ToNamespace, ValueType};
use crate::xotdata::{Node, Xot};

// pretty printing

// approach: modify tree to insert whitespace
// alternative approach: modify serializer to insert whitespace

// modify serializer rule:
// when we insert a tag open, we put in the indentation level,
// unless this tag is part of a mixed content node
// when we insert a tag close, we reduce the indentation level
// each tag is inserted. If we insert any other node, we don't mess with indentation.

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum StackEntry {
    Unmixed,
    Mixed,
}

struct PrettyWriter<'a, W: Write> {
    xot: &'a Xot<'a>,
    inner_writer: XmlSerializerWriter<'a, W>,
    stack: Vec<StackEntry>,
}

impl<'a, W: Write> PrettyWriter<'a, W> {
    fn new(xot: &'a Xot<'a>, w: W) -> PrettyWriter<'a, W> {
        let inner_writer = XmlSerializerWriter::new(xot, w);
        PrettyWriter {
            xot,
            inner_writer,
            stack: Vec::new(),
        }
    }

    fn unmixed(&mut self) {
        self.stack.push(StackEntry::Unmixed);
    }

    fn mixed(&mut self) {
        self.stack.push(StackEntry::Mixed);
    }

    fn in_mixed(&self) -> bool {
        self.stack.iter().any(|e| *e == StackEntry::Mixed)
    }

    fn pop(&mut self) {
        self.stack.pop();
    }

    fn write_indentation(&mut self) -> Result<(), Error> {
        if self.in_mixed() {
            return Ok(());
        }
        let indent = self
            .stack
            .iter()
            .filter(|e| **e == StackEntry::Unmixed)
            .count();
        let indent = " ".repeat(indent * 2);
        self.inner_writer.write(&indent)?;
        Ok(())
    }

    fn write_newline(&mut self) -> Result<(), Error> {
        self.inner_writer.write("\n")?;
        Ok(())
    }

    fn indent(&mut self, node: Node) -> Result<(), Error> {
        let has_children = self.xot.first_child(node).is_some();
        if has_children {
            if !self.has_text_child(node) {
                self.write_newline()?;
                self.unmixed();
            } else {
                self.mixed();
            }
        }
        Ok(())
    }

    fn dedent(&mut self, node: Node) -> Result<(), Error> {
        let has_children = self.xot.first_child(node).is_some();
        if has_children {
            let was_in_mixed = self.in_mixed();
            self.pop();
            if !was_in_mixed {
                self.write_indentation()?;
            }
        }
        Ok(())
    }

    fn dedent_newline(&mut self) -> Result<(), Error> {
        if !self.in_mixed() {
            self.write_newline()?;
        }
        Ok(())
    }

    fn has_text_child(&self, node: Node) -> bool {
        self.xot
            .children(node)
            .any(|child| self.xot.value_type(child) == ValueType::Text)
    }
}

impl<'a, W: Write> SerializerWriter for PrettyWriter<'a, W> {
    fn fullname(&self, name_id: NameId) -> Result<String, Error> {
        self.inner_writer.fullname(name_id)
    }

    fn fullname_attr(&self, name_id: NameId) -> Result<String, Error> {
        self.inner_writer.fullname_attr(name_id)
    }

    fn push_prefixes(&mut self, prefixes: &ToNamespace) {
        self.inner_writer.push_prefixes(prefixes);
    }

    fn pop_prefixes(&mut self, prefixes: &ToNamespace) {
        self.inner_writer.pop_prefixes(prefixes);
    }

    fn write_start_tag_open(&mut self, node: Node, element: &Element) -> Result<(), Error> {
        self.write_indentation()?;
        self.inner_writer.write_start_tag_open(node, element)
    }

    fn write_start_tag_close(&mut self, node: Node, element: &Element) -> Result<(), Error> {
        self.inner_writer.write_start_tag_close(node, element)?;

        self.indent(node)?;
        Ok(())
    }

    fn write_end_tag(&mut self, node: Node, element: &Element) -> Result<(), Error> {
        self.dedent(node)?;
        self.inner_writer.write_end_tag(node, element)?;
        self.dedent_newline()?;
        Ok(())
    }

    fn write_namespace_declaration(
        &mut self,
        node: Node,
        element: &Element,
        prefix_id: PrefixId,
        namespace_id: NamespaceId,
    ) -> Result<(), Error> {
        self.inner_writer
            .write_namespace_declaration(node, element, prefix_id, namespace_id)?;
        Ok(())
    }

    fn write_attribute(
        &mut self,
        node: Node,
        element: &Element,
        name_id: NameId,
        value: &str,
    ) -> Result<(), Error> {
        self.inner_writer
            .write_attribute(node, element, name_id, value)?;
        Ok(())
    }

    fn write_text(&mut self, node: Node, text: &Text) -> Result<(), Error> {
        self.inner_writer.write_text(node, text)?;
        Ok(())
    }

    fn write_comment(&mut self, node: Node, comment: &Comment) -> Result<(), Error> {
        self.write_indentation()?;
        self.inner_writer.write_comment(node, comment)?;
        self.dedent_newline()?;
        Ok(())
    }

    fn write_processing_instruction(
        &mut self,
        node: Node,
        pi: &ProcessingInstruction,
    ) -> Result<(), Error> {
        self.write_indentation()?;
        self.inner_writer.write_processing_instruction(node, pi)?;
        self.dedent_newline()?;
        Ok(())
    }

    fn write_space(&mut self) -> Result<(), Error> {
        self.inner_writer.write_space()?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;

    #[test]
    fn test_only_elements() {
        let mut xot = Xot::new();

        let doc = xot.parse(r#"<doc><a><b/></a></doc>"#).unwrap();

        let mut buf = Vec::new();
        let mut writer = PrettyWriter::new(&xot, &mut buf);
        xot.serialize_with_writer(doc, &mut writer).unwrap();
        let s = String::from_utf8(buf).unwrap();
        assert_snapshot!(s);
    }

    #[test]
    fn test_only_elements_multi() {
        let mut xot = Xot::new();

        let doc = xot
            .parse(r#"<doc><a><b/></a><a><b/><b/></a></doc>"#)
            .unwrap();

        let mut buf = Vec::new();
        let mut writer = PrettyWriter::new(&xot, &mut buf);
        xot.serialize_with_writer(doc, &mut writer).unwrap();
        let s = String::from_utf8(buf).unwrap();
        assert_snapshot!(s);
    }

    #[test]
    fn test_text() {
        let mut xot = Xot::new();

        let doc = xot.parse(r#"<doc><a>text</a><a>text 2</a></doc>"#).unwrap();

        let mut buf = Vec::new();
        let mut writer = PrettyWriter::new(&xot, &mut buf);
        xot.serialize_with_writer(doc, &mut writer).unwrap();
        let s = String::from_utf8(buf).unwrap();
        assert_snapshot!(s);
    }

    #[test]
    fn test_comment() {
        let mut xot = Xot::new();

        let doc = xot
            .parse(r#"<doc><a><!--hello--><!--world--></a></doc>"#)
            .unwrap();

        let mut buf = Vec::new();
        let mut writer = PrettyWriter::new(&xot, &mut buf);
        xot.serialize_with_writer(doc, &mut writer).unwrap();
        let s = String::from_utf8(buf).unwrap();
        assert_snapshot!(s);
    }

    #[test]
    fn test_pi() {
        let mut xot = Xot::new();
        let doc = xot.parse(r#"<doc><a><?pi?><?pi?></a></doc>"#).unwrap();
        let mut buf = Vec::new();
        let mut writer = PrettyWriter::new(&xot, &mut buf);
        xot.serialize_with_writer(doc, &mut writer).unwrap();
        let s = String::from_utf8(buf).unwrap();
        assert_snapshot!(s);
    }
}
