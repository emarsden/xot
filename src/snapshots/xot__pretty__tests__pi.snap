---
source: src/pretty.rs
expression: s
---
<doc>
  <a>
    <?pi?>
    <?pi?>
  </a>
</doc>

